tests/test_optimize.py

```diff
@@ -1,13 +1,68 @@
+import timemachine
+
 from optimize.step import truncated_step
+from optimize.utils import flatten_and_unflatten
+
+from ff import Forcefield
+from ff.handlers.deserialize import deserialize_handlers
+
+from pathlib import Path
 
 import numpy as onp
 
-from jax import numpy as jnp
-from jax import grad
-from jax.config import config
+from jax import tree_util, grad, config, numpy as jnp
+
 config.update("jax_enable_x64", True)
 
 
+def check_flatten_and_unflatten_roundtrip(example_pytree):
+    """assert that structures with array-valued leaves can be successfully
+    flattened and unflattened"""
+
+    flatten, unflatten = flatten_and_unflatten(example_pytree)
+
+    original_vector = flatten(example_pytree)
+    roundtripped_vector = flatten(unflatten(original_vector))
+
+    assert (original_vector == roundtripped_vector).all()
+
+    original_structure = tree_util.tree_structure(example_pytree)
+    roundtripped_structure = tree_util.tree_structure(unflatten(flatten(example_pytree)))
+
+    assert original_structure == roundtripped_structure
+
+
+def test_flatten_and_unflatten_dict():
+    """flatten/unflatten a nested object with array-valued leaves"""
+
+    example_pytree = dict(
+        Bonds=onp.random.randn(50, 2),
+        Angles=onp.random.randn(60, 3),
+        Extras=dict(
+            GlobalVars=[
+                jnp.arange(70),
+                onp.random.randn(80, 4)
+            ]
+        )
```

Comment:
Possible to also test a scalar?
eg. Bias=1.2, Multiplier=2.5

Reply:
This assumes arrays -- would `Bias=np.array([1.2])` be okay to add for this test or should I modify flatten_and_unflatten to handle float-valued as well as array-valued leaves?

Reply:
Added length-one arrays to this test in https://github.com/proteneer/timemachine/pull/474/commits/97ec17aed72058f08831e7183269d03267662860 .

Modifying flatten_and_unflatten to handle float-valued as well as array-valued leaves shortly...

Reply:
Float-valued leaves handled in https://github.com/proteneer/timemachine/pull/474/commits/9d258e34c667640f7afc3fed38d52aca24c0a6bf